File: 01_GOF23/DP_examples/C/03_Behavioral/State/context.c
#include "context.h"
#include "concrete_state.h"

// 初始化TCPConnection
void TCPConnection_Init(TCPConnection* connection) {
    connection->_tcpState = (TCPState*)GetTCPUnconnectedInstance();
    connection->Connector = INVALID_SOCKET;
    strcpy(connection->IPAddress, "127.0.0.1");
    connection->Port = 7777;
}

// 连接方法
void TCPConnection_Connect(TCPConnection* connection) {
    connection->_tcpState->Connect(connection->_tcpState, connection);
}

// 接收方法
void TCPConnection_Receive(TCPConnection* connection, char** buffer, int* buffer_len) {
    connection->_tcpState->Receive(connection->_tcpState, connection, buffer, buffer_len);
}

// 关闭方法
void TCPConnection_Close(TCPConnection* connection) {
    connection->_tcpState->Close(connection->_tcpState, connection);
}

// 发送消息方法
void TCPConnection_SendMessage(TCPConnection* connection, char* buffer, int buffer_len) {
    connection->_tcpState->SendMessage(connection->_tcpState, connection, buffer, buffer_len);
}

// 更改状态方法
void TCPConnection_ChangeState(TCPConnection* connection, TCPState* state) {
    connection->_tcpState = state;
}

// 检查连接状态
int TCPConnection_IsConnected(TCPConnection* connection) {
    if (connection->Connector != INVALID_SOCKET) {
        // 在Windows上，我们可以使用WSAIoctl来检查连接状态
        // 这里简化处理，返回true表示已连接
        return 1;
    }
    return 0;
}

File: 01_GOF23/DP_examples/Rust/03_Behavioral/State/context.rs
use std::net::TcpStream;
use std::sync::Arc;

use crate::concrete_state::TCPUnconnected;

// 状态接口
pub trait TCPState {
    fn connect(&self, connection: &mut TCPConnection);
    fn receive(&self, connection: &mut TCPConnection) -> Option<Vec<u8>>;
    fn close(&self, connection: &mut TCPConnection);
    fn send_message(&self, connection: &mut TCPConnection, buffer: &[u8]);
    fn change_state(&self, connection: &mut TCPConnection, state: Arc<dyn TCPState>);
}

// 上下文结构体
pub struct  TCPConnection {
    tcp_state: Arc<dyn TCPState>,
    pub connector: Option<TcpStream>,
    ip_address: String,
    port: u16,
}

impl TCPConnection {
    pub fn new() -> Self {
        Self {
            tcp_state: Arc::new(TCPUnconnected::new()),
            connector: None,
            ip_address: "127.0.0.1".to_string(),
            port: 7777,
        }
    }

    pub fn connect(&mut self) {
        let state = Arc::clone(&self.tcp_state);
        state.connect(self);
    }

    pub fn receive(&mut self) -> Option<Vec<u8>> {
        let state = Arc::clone(&self.tcp_state);
        state.receive(self)
    }

    pub fn close(&mut self) {
        let state = Arc::clone(&self.tcp_state);
        state.close(self);
    }

    pub fn send_message(&mut self, buffer: &[u8]) {
        let state = Arc::clone(&self.tcp_state);
        state.send_message(self, buffer);
    }

    pub fn is_connected(&self) -> bool {
        self.connector.is_some()
    }

    pub fn get_connector(&mut self) -> Option<&mut TcpStream> {
        self.connector.as_mut()
    }

    pub fn set_connector(&mut self, connector: TcpStream) {
        self.connector = Some(connector);
    }

    pub fn clear_connector(&mut self) {
        self.connector = None;
    }

    pub fn get_ip_address(&self) -> &str {
        &self.ip_address
    }

    pub fn get_port(&self) -> u16 {
        self.port
    }

    pub fn change_state(&mut self, state: Arc<dyn TCPState>) {
        self.tcp_state = state;
    }
}
